iavl/iavl_proof2.go

```diff
@@ -0,0 +1,251 @@
+	if !bytes.Equal(proof.RootHash, root) {
+		return errors.New("roots do not match")
+	}
+
+	if proof.LeftPath == nil && proof.RightPath == nil {
+		return errors.New("at least one path must exist")
+	}
+	if proof.LeftPath != nil {
+		if err := proof.LeftPath.Verify(proof.LeftNode, root); err != nil {
+			return errors.New("failed to verify left path")
+		}
+		if bytes.Compare(proof.LeftNode.KeyBytes, key) != -1 {
+			return errors.New("left node key must be lesser than supplied key")
+		}
+	}
+	if proof.RightPath != nil {
+		if err := proof.RightPath.Verify(proof.RightNode, root); err != nil {
+			return errors.New("failed to verify right path")
+		}
+		if bytes.Compare(proof.RightNode.KeyBytes, key) != 1 {
+			return errors.New("left node key must be greater than supplied key")
+		}
+	}
+
+	// Both paths exist, check that they are sequential.
+	if proof.RightPath != nil && proof.LeftPath != nil {
+		// TODO
+		return nil
+	}
+
+	// Only right path exists, check that node is at left boundary.
+	if proof.LeftPath == nil {
+		if !proof.RightPath.IsLeftmost() {
+			return errors.New("right path is only one but not leftmost")
+		}
+	}
+
+	// Only left path exists, check that node is at right boundary.
+	if proof.RightPath == nil {
+		if !proof.LeftPath.IsRightmost() {
+			return errors.New("left path is only one but not rightmost")
+		}
+	}
+
+	return nil
+}
+
+type KeyRangeExistsProof struct {
+	RootHash   []byte
+	PathToKeys []*PathToKey
+}
+
+func (proof *KeyRangeExistsProof) Verify(key []byte, value []byte, root []byte) bool {
+	return false
+}
+
+func ReadKeyExistsProof(data []byte) (*KeyExistsProof, error) {
+	proof := new(KeyExistsProof)
+	err := wire.ReadBinaryBytes(data, &proof)
+	if err != nil {
+		return nil, err
+	}
+	return proof, nil
+}
+
+func (node *IAVLNode) constructKeyExistsProof(t *IAVLTree, key []byte, proof *KeyExistsProof) (value []byte, exists bool) {
+	if node.height == 0 {
+		if bytes.Compare(node.key, key) == 0 {
+			proof.LeafHash = node.hash
+			return node.value, true
+		}
+		return nil, false
+	}
+
+	if bytes.Compare(key, node.key) < 0 {
+		if value, exists := node.getLeftNode(t).constructKeyExistsProof(t, key, proof); exists {
+			branch := IAVLProofInnerNode{
+				Height: node.height,
+				Size:   node.size,
+				Left:   nil,
+				Right:  node.getRightNode(t).hash,
+			}
+			proof.InnerNodes = append(proof.InnerNodes, branch)
+			return value, true
+		}
+		return nil, false
+	}
+
+	if value, exists = node.getRightNode(t).constructKeyExistsProof(t, key, proof); exists {
+		branch := IAVLProofInnerNode{
+			Height: node.height,
+			Size:   node.size,
+			Left:   node.getLeftNode(t).hash,
+			Right:  nil,
+		}
+		proof.InnerNodes = append(proof.InnerNodes, branch)
+		return value, true
+	}
+	return nil, false
+}
+
+func (node *IAVLNode) constructKeyNotExistsProof(t *IAVLTree, key []byte, proof *KeyNotExistsProof) error {
+	idx, _, exists := t.Get(key)
```

Comment:
Please add comment that idx is the index of the first key greater than the requested key if it doesn't exist (or max+1 if higher than all)